OpenSim/Simulation/StatesTrajectory.cpp

```diff
@@ -0,0 +1,393 @@
+            NonUniqueColumnsInStatesStorage);
+
+    // To be safe, we don't process storages in which individual rows are
+    // missing values.
+    OPENSIM_THROW_IF(numDependentColumns != sto.getSmallestNumberOfStates(),
+            VaryingNumberOfStatesPerRow,
+            numDependentColumns, sto.getSmallestNumberOfStates());
+
+    // Check if states are missing from the Storage.
+    // ---------------------------------------------
+    const auto& modelStateNames = localModel.getStateVariableNames();
+    std::vector<std::string> missingColumnNames;
+    // Also, assemble the names of the states that we will actually set in the
+    // trajectory, along with the corresponding state index.
+    std::map<int, std::string> statesToFillUp;
+    for (int is = 0; is < modelStateNames.getSize(); ++is) {
+        // getStateIndex() will check for pre-4.0 column names.
+        const int stateIndex = sto.getStateIndex(modelStateNames[is]);
+        if (stateIndex == -1) {
+            missingColumnNames.push_back(modelStateNames[is]);
+        } else {
+            statesToFillUp[stateIndex] = modelStateNames[is];
+        }
+    }
+    OPENSIM_THROW_IF(!allowMissingColumns && !missingColumnNames.empty(),
+            MissingColumnsInStatesStorage, 
+            localModel.getName(), missingColumnNames);
+
+    // Check if the Storage has columns that are not states in the Model.
+    // ------------------------------------------------------------------
+    if (!allowExtraColumns) {
+        if (numDependentColumns > statesToFillUp.size()) {
+            std::vector<std::string> extraColumnNames;
+            // We want the actual column names, not the state names; the two
+            // might be different b/c the state names changed in v4.0.
+            for (int ic = 1; ic < stoLabels.getSize(); ++ic) {
+                // Has this label been marked as a model state?
+                // stateIndex = columnIndex - 1 (b/c of the time column).
+                if (statesToFillUp.count(ic - 1) == 0) {
+                    extraColumnNames.push_back(stoLabels[ic]);
+                }
+            }
+            OPENSIM_THROW(ExtraColumnsInStatesStorage, localModel.getName(),
+                    extraColumnNames);
+        }
+    }
+
+
+    // Fill up trajectory.
+    // ===================
+
+    // Reserve the memory we'll need to fit all the states.
+    states.m_states.reserve(sto.getSize());
+
+    // Working memory.
+    SimTK::Vector dependentValues(numDependentColumns);
+
+    // Loop through all rows of the Storage.
+    for (int itime = 0; itime < sto.getSize(); ++itime) {
+
+        // Extract data from Storage into the working Vector.
+        sto.getData(itime, numDependentColumns, dependentValues);
+
+        // Set the correct time in the state.
+        sto.getTime(itime, state.updTime());
+
+        // Put in NaNs for state variable values not in the Storage.
+        for (const auto& stateName : missingColumnNames) {
```

Comment:
Should this be NaN or model's default for this state value? NaN may be better so that the value is not blinded used as though it were read in - when in fact it did not exist. A StateTrajectoryBuilder could read in a StateTrajectory and produce a new with NaNs resolved based on convenient rules, like use default values and solver for equilibrium, etc ... is that what you are thinking?